### PR TITLE
GH-12: Add possibility to run daemon in background and control it

### PR DESCRIPTION
**Pull request by maruu**
This is a draft implementation using a new `control` service to control the shutdown of the daemon.
The actual shutdown of the daemon (`os.Exit()`) can (and will) be further improved if the approach in general is agreed upon.

### CODE REVIEW

File: cmd/daemon.go [hunk 1]
```diff
@@ -15,14 +21,62 @@ func NewDaemonCmd(global *globalFlags) *cobra.Command {
 		Use:   "daemon",
 		Short: "Run the crypta daemon",
 		RunE: func(cmd *cobra.Command, args []string) error {
-			return c.Run(args)
+			if !c.detached {
+				return cmd.Help()
+			}
```
Comment (codetent) [inline cmd/daemon.go:26]: Since the behavior of the command is different according to the flag, I would consider creating a separate subcommand like `run` and convert the `daemon` to just the group of the other three subcommands.

File: cmd/daemon.go [hunk 2]
```diff
@@ -15,14 +21,62 @@ func NewDaemonCmd(global *globalFlags) *cobra.Command {
...
+	cc.AddCommand(stop)
+
+	cc.Flags().BoolVar(&c.detached, "detached", false, "Runs the daemon detached")
+	_ = cc.Flags().MarkHidden("detached")
+
 	return cc
 }
 
-func (c *daemonCmd) Run(args []string) error {
+func (c *daemonCmd) start(args []string) error {
```
Comment (codetent) [inline cmd/daemon.go:57]: I would not combine these three independent entry points in the `daemonCmd` struct. Instead I would split it up by responsibility: one for running, one for starting and one for stopping the daemon. This would also allow us to have seperate flagsets for each subcommand.

File: cmd/daemon.go [hunk 3]
```diff
@@ -15,14 +21,62 @@ func NewDaemonCmd(global *globalFlags) *cobra.Command {
...
+	if err != nil {
+		return err
+	}
+	ex, err := os.Executable()
+	if err != nil {
+		return err
+	}
+
+	// FIXME: Pass global flags to command, in order to be able to set the endpoint
+	args = append(args, "daemon", "--detached")
```
Comment (codetent) [inline cmd/daemon.go:68]: Why do you append the `daemon` subcommand to the positional args passed to the `start` command?

File: control/v1/control.proto [hunk 4]
```diff
@@ -0,0 +1,15 @@
+syntax = "proto3";
+
+package control.v1;
+
+option go_package = "example/gen/control/v1;controlv1";
```
Comment (codetent) [inline control/v1/control.proto:5]: Why is this needed?

File: pkg/daemon/client.go [hunk 5]
```diff
@@ -16,20 +18,26 @@ var (
...
 func NewDaemonClient(ip string, port string) *daemonClient {
+	endpoint := fmt.Sprintf("http://%s:%s", ip, port)
 	return &daemonClient{
-		client: secretv1connect.NewSecretServiceClient(
+		secretClient: secretv1connect.NewSecretServiceClient(
 			http.DefaultClient,
-			fmt.Sprintf("http://%s:%s", ip, port),
+			endpoint,
+		),
+		controlClient: controlv1connect.NewControlServiceClient(
```
Comment (codetent) [inline pkg/daemon/client.go:32]: I like the concept of having the services separated by their responsibility but I'm not sure if it is worth having two grpc connections in one client instance especially if we probably just use one at a time.
I we split the service, I would also have a different client wrappers.

In general I'm not sure if the `Shutdown` command is worth the overhead of having a separate client and grpc connection or if it is better to keep everything in a `DaemonService` for now and split it in the future if really needed.

File: pkg/daemon/controlhandler.go [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+func NewControlServiceHandler(shut shutdowner) (string, http.Handler) {
+	return controlv1connect.NewControlServiceHandler(&controlServiceServer{shutdown: shut})
+}
+
+func (s *controlServiceServer) Shutdown(
+	ctx context.Context,
+	req *connect.Request[controlv1.ShutdownRequest],
+) (*connect.Response[controlv1.ShutdownResponse], error) {
+	// wait so the response can be returned, and the server is shutdown afterwards
+	go time.AfterFunc(1*time.Second, s.shutdown)
```
[the remark this answers is not in the document]
Comment (codetent) [reply pkg/daemon/controlhandler.go:29]: H'm it looks a bit errorprone because it can cause some race conditions.
This concept has some potential issues:
- Server shuts down and response is not sent (e.g. if message is not sent immediately)
- Response is sent but server does not shutdown because of any reason
- Client does not know either way if the shutdown was successful
Not sure if there is anything we can do other than passing waiting for the message to go out, and then passing the control to the shutdown function after closing the grpc server. Unfortunately also in this case the client has no ACK if it really happened.

What we could do instead:
- The client asks the server for its PID over grpc
- The client sends a signal like SIGTERM to the server
- The server catches the signal and shuts down gracefully
- The client checks the pid with a timeout to have an ACK

What do you think?

File: cmd/daemon.go (hunk 1 above)
Comment (maruu) [reply cmd/daemon.go:26]: The "nice thing" about this implementation is that the flag is hidden, and the daemon "main command" just returns the help message of the two sub commands if called without the flag. This way it is clear that the daemon needs to be started and stopped using the two sub commands.

Of course, if we want to support a separate `run` command that should also be executable by the user as well (i.e. not hidden), it makes sense to have it visible. But usually I would prefer to only have one way to start/stop something as it makes the interface easier to understand. Supporting both ways (asynchronous & synchronous) could lead to confusion I think. ("Do I need to call "run" after calling "start"?)

File: cmd/daemon.go (hunk 2 above)
Comment (maruu) [reply cmd/daemon.go:57]: So you would propose to introduce 3 structs for the 3 entry points?

I am not sure about this. If we would like to change the flagset in the future, we could still do this change. Until then, I would propose to follow `YAGNI` and keep it as simple as possible.

I currently don't see any need for different flagsets for the three "commands". Do you?

File: pkg/daemon/client.go (hunk 5 above)
Comment (maruu) [reply pkg/daemon/client.go:32]: I am a little bit torn.

On the one hand, you are right. The new service adds overhead & complexity. It would only make sense to make this change now if we could already think about adding additional "control" functionality. I personally cannot think of any at the moment, apart from the `Shutdown`.

On the other hand, the separation of concerns and single responsibility patterns tell me that its probably not a good idea to mix the secret storage with control mechanisms. 

In general I would prefer to keep them separate, but I admit that this would only make sense in the future if we were to add more "control" functionality. I will create a single DaemonService instead.

File: control/v1/control.proto (hunk 4 above)
Comment (maruu) [reply control/v1/control.proto:5]: Will be removed by the merge to the `DaemonService` :-)

File: cmd/daemon.go (hunk 3 above)
Comment (maruu) [reply cmd/daemon.go:68]: I just reused the `args` array for passing the new arguments to the new subprocess (as they are empty anyway for the daemon command). This will not be this way in the final PR.

File: pkg/daemon/controlhandler.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (maruu) [reply pkg/daemon/controlhandler.go:29]: Yes, good idea. I will try it out and update the PR.

[the remark this answers is not in the document]
Comment (maruu) [reply pkg/daemon/controlhandler.go:29]: The problem I see with the approach is that it also contains raceconditions. For example:

Between 
- The server catches the signal and shuts down gracefully
- The client checks the pid with a timeout to have an ACK

The server might have already shutdown before the client checks the PID, and the PID might have been reused by the system for a new process. The client would therefore observe that the process is still alive, and has not been shutdown.

File: cmd/daemon.go (hunk 2 above)
Comment (codetent) [reply cmd/daemon.go]: Yes I would keep the structure consistent and have one flagset and one entrypoint per struct like for the other commands. In this case the struct is a container for a command logic with its flags. Since these logics and the corresponding flags can change indepedently for different reasons at different times, I would stick to the SRP and separate them.

Regarding the flagsets I agree. In the future it could be that some arguments for the automatic stopping of the daemon could be added to `start` and a timeout for killing it the to `stop` command. For now I would also keep it as simple as possible.

File: cmd/daemon.go (hunk 1 above)
Comment (codetent) [reply cmd/daemon.go:32]: I agree regarding the interface to keep it simple now.

My initial idea was to create a hidden run command that only runs the daemon to keep responsibilities clear. This looks a little bit like a code smell because a boolean variable is used to switch between to different inner logics (running daemon, showing help page). Or in other words: if the flag is present it behaves like a group, otherwise like a command.

I'm not sure if cobra supports such hidden commands. Otherwise we could also keep it like that for now.

File: pkg/daemon/controlhandler.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (codetent) [reply pkg/daemon/controlhandler.go]: Yes you are absolutely right. This could happen but I guess this can only happen if either it takes a lot of time in between or there are a lot of processes. I think it is pretty unlikely that it will directly reassign it.

Maybe we could inspect the process and store some information about it (e.g. command line, uptime or similar) before sending the signal. Then we could check again if it's the same process instead of just checking the pid.

There is also other potential race conditions between receiving the PID from the server, sending the kill signal and actually shutting down the server. If the endpoint would also return the other information (maybe as a hash to compare) we could also use it for identification later but in the end I guess you cannot completly get rid of all race conditions. We could just reduce the probability.

**Comment by maruu** [comment]
Analysis update:

- It seems that the daemon is not stopped, even if os.Exit(0) is called directly in the daemon. The process is still shown in the list of running processes.

**Comment by maruu** [comment]
PR is closed because it does not seem to be possible to detach a process, and have it cleaned up by the init process in docker containers, especially if the init process does not check the exit status of their child processes (as is the case, e.g. for github container runners, which use tail -f as the entry point (init process)).

Crypta will draw its system boundary at the actual startup of the daemon & the CLI itself. The user of crypta can use onboard tools specific for their environment to handle the usecase of running the daemon in the background.
